mailbox/tools/indexer/src/main/java/org/apache/mailbox/tools/indexer/ReIndexerPerformer.java

```diff
@@ -186,30 +202,54 @@ public ReIndexerPerformer(MailboxManager mailboxManager,
-                .map(message -> new ReIndexingEntry(mailbox, mailboxSession, message)));
+                .map(message -> new ReIndexingEntry(mailbox, mailboxSession, message)))
+            .onErrorResume(e -> {
+                LOGGER.warn("ReIndexing failed for {}", failure, e);
+                reprocessingContext.recordFailureDetailsForMessage(failure.getMailboxId(), failure.getUid());
+                return Mono.empty();
+            });
     }
 
-    private Flux<ReIndexingEntry> reIndexingEntriesForMailbox(Mailbox mailbox, MailboxSession mailboxSession) {
+    private Flux<ReIndexingEntry> reIndexingEntriesForMailbox(Mailbox mailbox, MailboxSession mailboxSession, ReprocessingContext reprocessingContext) {
         MessageMapper messageMapper = mailboxSessionMapperFactory.getMessageMapper(mailboxSession);
 
         return messageSearchIndex.deleteAll(mailboxSession, mailbox.getMailboxId())
             .thenMany(messageMapper.listAllMessageUids(mailbox))
-            .flatMap(uid -> messageMapper.findInMailboxReactive(mailbox, MessageRange.one(uid), MessageMapper.FetchType.Full, UNLIMITED))
-            .map(message -> new ReIndexingEntry(mailbox, mailboxSession, message));
+            .flatMap(uid -> reIndexingEntryForUid(mailbox, mailboxSession, reprocessingContext, messageMapper, uid))
+            .onErrorResume(e -> {
+                LOGGER.warn("ReIndexing failed for {}", mailbox.generateAssociatedPath(), e);
+                reprocessingContext.recordMailboxFailure(mailbox.getMailboxId());
+                return Mono.empty();
+            });
+    }
+
+    private Flux<ReIndexingEntry> reIndexingEntryForUid(Mailbox mailbox, MailboxSession mailboxSession, ReprocessingContext reprocessingContext, MessageMapper messageMapper, MessageUid uid) {
+        return messageMapper.findInMailboxReactive(mailbox, MessageRange.one(uid), MessageMapper.FetchType.Full, UNLIMITED)
+            .map(message -> new ReIndexingEntry(mailbox, mailboxSession, message))
+            .onErrorResume(e -> {
+                LOGGER.warn("ReIndexing failed for {} - {}", mailbox.getMailboxId(), uid, e);
+                reprocessingContext.recordFailureDetailsForMessage(mailbox.getMailboxId(), uid);
+                return Mono.empty();
+            });
     }
 
     private Mono<Task.Result> reIndexMessages(Flux<ReIndexingEntry> entriesToIndex, RunningOptions runningOptions, ReprocessingContext reprocessingContext) {
         return ReactorUtils.Throttler.<ReIndexingEntry, Task.Result>forOperation(entry -> reIndexMessage(entry.getMailboxSession(), entry.getMailbox(), reprocessingContext, entry.getMessage()))
             .window(runningOptions.getMessagesPerSecond(), Duration.ofSeconds(1))
             .throttle(entriesToIndex)
             .reduce(Task::combine)
-            .switchIfEmpty(Mono.just(Result.COMPLETED));
+            .switchIfEmpty(Mono.fromSupplier(() -> {
+                if (reprocessingContext.failures().mailboxFailures().isEmpty()) {
```

Comment:
I would like not to rely on side effects in order to determine if we are a success or a failure.

Isn't it handled by the combine the line above?

Reply:
We are reading mailbox(es) in order to retrieve a Flux of all messages (for throttling stuff) then we pass it to the throttling function. If an error occur when reading message (at mailbox level), we return a Mono.empty() in the mapper & record 1 mailbox failure.

The problem is after that if we don't consider the content of mailboxFailures then there's no way to detect the case where there were mailbox failures (PARTIAL result) and other cases where there were not, for example an empty mailbox.

May be there is a better way to handle this because this hack hurts my eyes also.

Reply:
I see.

> May be there is a better way to handle this because this hack hurts my eyes also.

Yes. Let's invest some time for that on monday.